### ISSUE TITLE
Corrupt PSD with Smart Objects

### ISSUE THREAD

**Issue by jeanpaze** [Agamnentzar/ag-psd#58]
Hey @Agamnentzar, thank you for this handy lib!

I'm having issues saving an existing PSD with Smart Object in it.

My main goal is to actually read a PSD (myPsd.psd inside this [psds.zip](https://github.com/Agamnentzar/ag-psd/files/6342924/psds.zip)), duplicate a specific layer group, and save it as a new one.

If I have a PSD that contains only text layers, it works. Even if I have an image (not as Smart Object) also works. When I include a Smart Object, the saved PSD is corrupt no matter the type of that Smart Object (PSB, JPG, etc).

I'm on macOS Catalina and my Node version is `10.19.0`, so I had to stick with ES5. I also tried to update to the latest Node version - faced the same issues.

This is my code:

```
const fs = require( 'fs-extra' );
const agPsd = require( 'ag-psd' );
require( 'ag-psd/initialize-canvas' );


// read document structure and image data
const bufferRead = fs.readFileSync( 'myPsd.psd' );

const psdRead = agPsd.readPsd( bufferRead, {
	skipLayerImageData:false,
	skipCompositeImageData:false,
	skipThumbnail:false,
	skipLinkedFilesData:false,
	throwForMissingFeatures:false,
	logMissingFeatures:true,
	useImageData:true,
	useRawThumbnail:true,
	logDevFeatures:true
} );

// manipulate psd
var txtTemplate = getTemplate( psdRead );
var psdWrite = appendTemplate( txtTemplate, psdRead, psdRead );

// write psd
const bufferWrite = agPsd.writePsdBuffer( psdWrite, {
	generateThumbnail:true,
	trimImageData:false,
	invalidateTextLayers:false,
	logMissingFeatures:true,
	noBackground:true
} );

fs.writeFileSync( 'myPsd_duplicated.psd', bufferWrite );


// Get template
function getTemplate( arr ){
	var list = arr.children || [];

	for( var i = 0; i < list.length; i++ ){
		if( list[ i ].name.indexOf( 'TXT-TEMPLATE' ) != -1 ){
			return list[ i ];
		} else {
			var result = getTemplate( list[ i ] );
			if( result ) return result;
		}
	};

	return null;
};

// Append template
function appendTemplate( template, target, originalPSD ){
	var list = target.children || [];

	for( var i = 0; i < list.length; i++ ){
		if( list[ i ].name.indexOf( 'TXT-TEMPLATE' ) != -1 ){
			list.splice( i, 0, template );

			return originalPSD;
		} else {
			var result = appendTemplate( template, list[ i ], target );
			if( result ) return result;
		}
	};

	return null;
};
```

When I run the code above and the PSD has a Smart Object, I get this (corrupted PSD):

```
Unread 3 bytes left for tag: PlLd
Unread 1 bytes left for tag: SoLd
Unhandled additional info: lnkE
Unhandled additional info: extn
```

If the PSD doesn't have a Smart Object, this happens (working PSD):

`Unhandled additional info: extn`

I ran `gulp cov` and this is the result:

```
$ gulp cov
Using gulpfile node_modules/ag-psd/gulpfile.js
Starting 'cov'...
Starting 'build-js'...
Finished 'build-js' after 7.59 s
Starting 'coverage'...


  ,․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․․
  ․․․․․․․․․․․․․․․․․․․Unhandled additional info: lnkE
․․․․․,,,,,,,,․․․․․․․․․․․․․․․․․․․․․․․․․․․․
  ․․

  113 passing (16s)
  9 pending

Finished 'coverage' after 17 s
Starting 'remap'...
Finished 'remap' after 453 ms
Finished 'cov' after 25 s
```

The test.psd is also attached.

Please let me know if you need more info regarding my environment.

Do you have any idea how can I get this to work?

Cheers!

**Comment by Agamnentzar** [comment]
I added few fixes for issues with reading your file and added automatic layer ID deduplication, which might have been an issue in your case, since it seems Photoshop is not dealing with duplicate layer IDs correctly, and you'll get duplicate IDs if you duplicate layers.

All of these fixes are released in version `11.6.2`

**Comment by jeanpaze** [comment]
That was fast! Thank you so much, I just tested and it's fixed now.

Cheers!
